src/math.cpp

```diff
@@ -240,51 +349,87 @@ auto fmod(const CallContext& ctx) -> Value {
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value, const Value& old_value1,
+                      const Value& old_value2) -> std::optional<SourceChangeTree> {
+            if (old_value2 == Nil()) {
+                Number n = std::get<Number>(new_value);
+                return old_value1.force(std::exp(n.value));
+            } else {
+                // log(x)/log(base) is ambiguously. so no clear return could be done
+                // if someone knows a way feel free to change it.
+                return std::nullopt;
+            }
+        }});
+
     auto x = ctx.arguments().get(0);
     auto base = ctx.arguments().get(1);
 
     return std::visit(
-        overloaded{
-            [ctx](auto /*unused*/, Nil /*unused*/) -> Value {
-                return math_helper(ctx, static_cast<double (*)(double)>(&std::log), "log");
-            },
-            [ctx](auto a, auto b) -> Value {
-                Vallist list({Value(a)});
-                CallContext new_ctx = ctx.make_new(list);
-                auto x = math_helper(new_ctx, static_cast<double (*)(double)>(&std::log), "log");
-                list = Vallist({Value(b)});
-                new_ctx = ctx.make_new(list);
-                auto base =
-                    math_helper(new_ctx, static_cast<double (*)(double)>(&std::log), "log", "2");
-                Number x_n = std::get<Number>(x);
-                Number base_n = std::get<Number>(base);
-                return Value(x_n.value / base_n.value);
-            }},
-        x.raw(), base.raw());
+               overloaded{
+                   [ctx](auto /*unused*/, Nil /*unused*/) -> Value {
+                       return math_helper(ctx, static_cast<double (*)(double)>(&std::log), "log");
+                   },
+                   [ctx](auto a, auto b) -> Value {
+                       Vallist list({Value(a)});
+                       CallContext new_ctx = ctx.make_new(list);
+                       auto x =
+                           math_helper(new_ctx, static_cast<double (*)(double)>(&std::log), "log");
+                       list = Vallist({Value(b)});
+                       new_ctx = ctx.make_new(list);
+                       auto base = math_helper(
+                           new_ctx, static_cast<double (*)(double)>(&std::log), "log", "2");
+                       Number x_n = std::get<Number>(x);
+                       Number base_n = std::get<Number>(base);
+                       return Value(x_n.value / base_n.value);
+                   }},
+               x.raw(), base.raw())
+        .with_origin(origin);
 }
 
 auto max(const CallContext& ctx) -> Value {
+    Origin origin = Origin(MultipleArgsOrigin{
+        .values = ctx.arguments(),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& /*new_value*/, const Vallist& /*args*/)
+            -> std::optional<SourceChangeTree> { return std::nullopt; }});
```

Comment:
Add a comment to say why we don't support forcing the result. I.e. because it is not obvious how to implement it if the value would be forced such that `max` would return a different value.

Same for `min`.